src/schemaVisitor.ts

```diff
@@ -0,0 +1,464 @@
+      // time a matching directive is found in the schema AST, with arguments
+      // and other metadata specific to that occurrence. To help prevent the
+      // mistake of passing instances, the SchemaDirectiveVisitor constructor
+      // method is marked as protected.
+      [directiveName: string]: typeof SchemaDirectiveVisitor
+    },
+    // Optional context object that will be available to all visitor instances
+    // via this.context. Defaults to an empty null-prototype object.
+    context: {
+      [key: string]: any
+    } = Object.create(null),
+  ): {
+    // The visitSchemaDirectives method returns a map from directive names to
+    // lists of SchemaDirectiveVisitors created while visiting the schema.
+    [directiveName: string]: SchemaDirectiveVisitor[],
+  } {
+    // If the schema declares any directives for public consumption, record
+    // them here so that we can properly coerce arguments when/if we encounter
+    // an occurrence of the directive while walking the schema below.
+    const declaredDirectives: {
+      [directiveName: string]: GraphQLDirective,
+    } = Object.create(null);
+    schema.getDirectives().forEach(decl => {
+      declaredDirectives[decl.name] = decl;
+    });
+
+    // Map from directive names to lists of SchemaDirectiveVisitor instances
+    // created while visiting the schema.
+    const createdVisitors: {
+      [directiveName: string]: SchemaDirectiveVisitor[]
+    } = Object.create(null);
+    Object.keys(directiveVisitors).forEach(directiveName => {
+      createdVisitors[directiveName] = [];
+    });
+
+    function visitorSelector(
+      type: VisitableSchemaType,
+      methodName: string,
+    ): SchemaDirectiveVisitor[] {
+      const visitors: SchemaDirectiveVisitor[] = [];
+      const directiveNodes = type.astNode && type.astNode.directives;
+      if (! directiveNodes) {
+        return visitors;
+      }
+
+      directiveNodes.forEach(directiveNode => {
+        const directiveName = directiveNode.name.value;
+        if (! hasOwn.call(directiveVisitors, directiveName)) {
+          return;
+        }
+
+        const visitorClass = directiveVisitors[directiveName];
+
+        // Avoid creating visitor objects if visitorClass does not override
+        // the visitor method named by methodName.
+        if (! visitorClass.implementsVisitorMethod(methodName)) {
+          return;
+        }
+
+        const decl = declaredDirectives[directiveName];
+        let args: { [key: string]: any };
+
+        if (decl) {
+          // If this directive was explicitly declared, use the declared
+          // argument types (and any default values) to check, coerce, and/or
+          // supply default values for the given arguments.
+          args = getArgumentValues(decl, directiveNode);
+        } else {
+          // If this directive was not explicitly declared, just convert the
+          // argument nodes to their corresponding JavaScript values.
```

Comment:
Note that the `SchemaDirectiveVisitor` abstraction works even for `@directive`s that have not been declared in the schema using the
```js
directive @myDirective(param: Type = defaultValue) on OBJECT | UNION | ...
```
syntax, though declaring your directives might still be a good idea, since the parameter types and default values will be enforced by this code.

Reply:
@stubailo Yes, if a `GraphQLDirective` node with the same name is in the schema at the time `visitSchemaDirectives` happens (usually at the end of `makeExecutableSchema`), then it will be enforced. Regarding locations, though, we only enforce that the `SchemaDirectiveVisitor` actually implements appropriate visitor methods for the locations declared by the declaration. It's ok if the visitor implements additional locations not mentioned by the declaration.